### DISCUSSION TITLE
Subtitles sometimes go out of sync

### DISCUSSION THREAD

**Discussion by athu16** [openai/whisper#89] [Q&A] [answered]
After generating a VTT file, the subtitles sometimes get hastened, and scroll by much faster than the actual audio. They usually get fixed within about 30 seconds (after one chunk of audio is processed), but if the audio is long, they start scrolling faster again after a short duration.
I've tried English, Russian and Marathi translation, and this problem usually occurs with Marathi which has a much higher WER %. I can understand the accuracy of translation being low, but I can't understand how a higher WER % could affect the timestamps.

**Comment by coder543** [comment]
I've personally observed this behavior more on the `large` model than on any of the other models.

**Reply by turnkit** [reply]
I had a 90 mp3 file go several minutes out of sync in large.  I had switched to small mp3 and the issues went away but I sure would like to return to the large model with confidence when a get a GPU.  I had assumed the issue was with mp3 time code tracking issue, so I hope it isn't somehow a large model problem.

**Comment by Xlindvain** [comment]
This is happening to me as well, using the Medium model, and generating a VTT for a video in Brazilian Portuguese.

**Comment by jongwook** [comment, marked as answer]
This is one of the failure mode of the hacky long-form heuristics (in [`transcribe.py`](https://github.com/openai/whisper/blob/main/whisper/transcribe.py) and discussed in Section 4.5), where the timestamp offsets sometimes accumulate over time, because the transcription from the previous 30-second window including the timestamps are fed to the model as conditioning input. This is currently controlled by a currently hard-coded constant here:

https://github.com/openai/whisper/blob/2d3032de014c514d5c029baccb898b5342463674/whisper/transcribe.py#L220-L222

and you can modify this block to always reset the context to mitigate the tendency of going out of sync.

In the paper, we have seen a slight improvement when using this previous-text conditioning than not:

<img width="516" alt="image" src="https://user-images.githubusercontent.com/266841/192140715-ebf3c489-d8d6-4876-910f-d05d7f54332a.png">

but it didn't always help (WER increased in TED-LIUM3). It might be worth making this previous-text conditioning as an optional flag, if the failure case is common in practice.

EDIT: just added `--condition_on_previous_text False` option. It may repeat short words across the boundaries but the timestamps should work better.

**Reply by athu16** [reply]
Thanks a lot! How do I update the package on my end? Do I just run the pip install command again?

**Reply by jltchiu** [reply]
@jongwook Do you think this conditioning input could be the cause for the different recognition result for two consecutive runs discussed in other thread?

**Reply by flesnuk** [reply]
@athu16 you can use the following command to force pip to get the latest version from the repo without fetching its dependencies:
`pip install --upgrade --no-deps --force-reinstall git+https://github.com/openai/whisper.git`

**Reply by jongwook** [reply]
@jltchiu Nondeterminism is caused by the temperature fallback heuristic which increases the temperature by 0.2 upon transcription failure; you can disable this heuristic by setting `--temperature_increment_on_fallback None`

**Reply by jltchiu** [reply]
Thanks, I am able to get same result after setting `--temperature_increment_on_fallback None`. However, I am curious why temperature parameter has different behavior with the same model and same audio file, mind to share the reason of it?

**Reply by flesnuk** [reply]
I myself don't know much about ASR terminology, but I found this small blog entry explaining what temperature does:
https://lukesalamone.github.io/posts/what-is-temperature/
So if temperature is 0, it always picks the option with the highest score, ignoring the other lower ranked tokens. Meanwhile incrementing the temperature it takes into account the other "less probable" tokens to be chosen, giving it the randomness nature you experience.

**Reply by jltchiu** [reply]
@flesnuk Thanks! This clarified my question! Really appreciate it.

**Reply by akashmjn** [reply]
@jongwook - appreciate the detailed response! Just to be super clear, in Table 7 - does this shows the impact of setting `--condition_on_previous_text=True` or is it the impact of making the heuristic intervention you linked?

https://github.com/openai/whisper/blob/2d3032de014c514d5c029baccb898b5342463674/whisper/transcribe.py#L220-L222

A bit confused by the complex semantics 😄 
_In the paper, we have seen a slight improvement when using this previous-text conditioning than not_
_Providing the transcribed text from the preceding window as previous-text conditioning when the applied temperature is below 0.5 further improves the performance_

Thanks!

**Reply by jongwook** [reply]
yes, after this thread I added the condition, and the three lines above now looks like this:

https://github.com/openai/whisper/blob/0b1ba3d46ebf7fe6f953acfd8cad62a4f851b49f/whisper/transcribe.py#L235-L237

The row "Previous-text conditioning" in Table 7 and below used the setting equivalent to `--condition_on_previous_text=True` (i.e. with the hard-coded 0.5 threshold), and the three rows above it used what's equivalent to `--condition_on_previous_text=False`.

**Reply by coder543** [reply]
As an interesting failure mode, attempting to transcribe [this](https://www.youtube.com/watch?v=DnnSKkd2etk) with the `large` model just resulted in the letter `a` getting printed repeatedly for over an hour of transcript. The `medium.en` model worked fine.

I found that setting `--temperature_increment_on_fallback None` sort of fixed the `large` model, and I wonder if this might have been caused by the 7 minutes of instrumental music at the beginning of the video, which might have caused repeated transcription failures.

Either way, I'm glad the `temperature_increment_on_fallback` option exists.

EDIT: on the whole, I think `medium.en` did better on this particular transcription anyways.

EDIT 2: It is amazing how much difference skipping the intro music makes for Whisper.

**Reply by dillfrescott** [reply]
I've also noticed that sometimes the smaller than large models seem to do just as good of a job if not better in many scenarios.

**Reply by ioudove** [reply]
> yes, after this thread I added the condition, and the three lines above now looks like this:
> 
> https://github.com/openai/whisper/blob/0b1ba3d46ebf7fe6f953acfd8cad62a4f851b49f/whisper/transcribe.py#L235-L237
> 
> The row "Previous-text conditioning" in Table 7 and below used the setting equivalent to `--condition_on_previous_text=True` (i.e. with the hard-coded 0.5 threshold), and the three rows above it used what's equivalent to `--condition_on_previous_text=False`.

Hi, I used to encounter this problem frequently, but after some testing, I found a solution.
When timestamps don’t match the subtitles, it’s usually caused by an issue with the audio file.
I used OBS to re-record the audio file, and that almost perfectly solved the problem.

However, if part of an audio file has no sound, it can still occasionally cause “phantom hearing” or timestamp errors.
When that happens, you can use **ffmpeg** to extract the problematic audio segment and transcribe it separately.
Then, adjust the starting timestamp of the separately transcribed subtitle file using code, and merge it back into the original subtitle file at the correct position — that will fix the issue.

**Tip for using OBS:**
To start “recording” and “playing the audio/video file” almost simultaneously, you can assign hotkeys — that way, both actions start at the same time.

Here’s the code I often use in Colab:
from google.colab import drive
drive.mount('/content/drive')

!pip install git+https://github.com/openai/whisper.git

!whisper "/content/drive/MyDrive/音频/1.aac" --model large --language zh --output_format srt --output_dir /content/drive/MyDrive/字幕/

!apt-get update && apt-get install -y ffmpeg

!ffmpeg -y -i "/content/drive/MyDrive/音频/重制转录完/1.wav" -ss 14160 -t 3600 -c copy /content/drive/MyDrive/音频/A/11-5.wav

!ffmpeg -y -i "/content/drive/MyDrive/音频/重制转录完/11.wav" -ss 17700 -c copy /content/drive/MyDrive/音频/A/1111.wav

!whisper /content/drive/MyDrive/音频/A/*.aac --model large --language zh --output_format srt --output_dir /content/drive/MyDrive/字幕

**Comment by rikhuijzer** [comment]
What also could be happening here is that the video that you run it on has gaps in the audio. For example, I was concatenating videos with FFmpeg like so
```sh
$ ffmpeg -f concat -i mylist.txt -c copy output.mp4
```
But if the concat list contains audio files and images, then it is possible that FFmpeg adds some padding in the video when going from one to the other for encoding reasons. During the padding, the audio can be put to zero. For example, here is an example which happened with an audio-only concat (https://superuser.com/questions/1600067/ffmpeg-concat-audio-without-spaces-between):

![image](https://github.com/user-attachments/assets/cee77764-e3e3-4abd-9b01-db1004453bdc)

But like I said this can happen with videos too.

The problem then is that when you point `whisper` to such a video file, it will use FFmpeg to convert the video to audio first (at least I'm 90% sure that's what happens, I didn't double check) and then this audio file can have a different length than the original video file because the audio drops the gaps. I've tested it myself that I had some video files with a duration of say 00:05:24.29 and then I ran
```sh
$ ffmpeg -i output.mp4 -q:a 0 -map a output.mp3
```
and suddenly the duration was shorter, say 00:05:14.59.

A way to fix it can be to use `-shorter` during the concat step inside FFmpeg when creating the videos. There is probably also a FFmpeg setting to not trim the pauses, but I haven't tried that.

**Reply by Tamalero** [reply]
Is there a way to force Whisper to use this command to prevent the gaps? I use Whisper extensively to translate for languages I do not understand  or cant get( Im deaf) but do not have available subs in websites.
And yes, its very accurate.. just the sync issues are very annoying.

**Reply by rikhuijzer** [reply]
Not that I know of, but in general you probably want to look into the C++ port of whisper https://github.com/ggml-org/whisper.cpp. It runs 4-8 times as fast on my laptop

**Reply by Tamalero** [reply]
> Not that I know of, but in general you probably want to look into the C++ port of whisper https://github.com/ggml-org/whisper.cpp. It runs 4-8 times as fast on my laptop

Thanks. LLMs suggested also to use STABLE-TS

**Comment by doited-boy** [comment]
I have the same issue with model "small" and the language "English" on Ubuntu 24.04. While subtitles itself are generally correct and incredibly accurate (kudos to the "Whisper" creators! you are awesome guys!), the timing is always terrible, and "out of sync" problem is very serious. So far I've made subtitles for short videos 10 minutes long and for long movies over 2 hours. The problem is always there. "Out of sync" text portions are everywhere, and they are heavy, too. Such errors occur every 20 minutes or so, then the timing gets even for some period, and then again we see the same errors.

It is very, very, VERY regrettable issue. Because you know, the "Whisper" is a wonderful and super useful application, and this "out of sync problem" is just a blight upon it's beautiful face.
